keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -0,0 +1,575 @@
+>>> default('{entry: null}')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{entry: {}}')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{entry: {name: other-name}}')
+{"entry": {"name": "other-name", "number": 0}}
+>>> default('{entry: {name: "", number: 0}}') # empty string is different from null or unspecified
+{"entry": {"name": "", "number": 0}}
+```
+
+On the other hand, this is forbidden:
+```golang
+Type Root struct {
+  // Defaults on non-pointer structs are FORBIDDEN:
+  // +default={name: entry, number: 12}
+  Entry SubLevel
+}
+
+Type SubLevel struct {
+  Name string
+  Number int
+}
+```
+
+#### Struct pointers
+
+Struct pointers are allowed to have a specific default since they are distinct
+from unspecified non-pointers structs. Modifying the type from the
+previous example:
+```golang
+type Root struct {
+  // This is now a pointer, we can provide a default:
+  // +default={name: pointer-name}
+  Entry *SubLevel
+}
+
+type SubLevel struct {
+  // +defaut=default-name
+  Name string
+  // +default=0
+  Number int
+}
+```
+
+We would generate the following OpenAPI definition, both for CRD types (by kubebuilder)
+and built-in types:
+```yaml
+Root:
+  type: object
+  default: {}
+  properties:
+    entry:
+      $ref: SubLevel
+      default: {name: pointer-name}
+SubLevel:
+  type: object
+  properties:
+    name:
+      type: string
+      default: default-name
+    number:
+      type: integer
+      default: 0
+```
+
+which would have the following behavior, both for CRD types and built-in types:
+```python
+>>> default('null')
+{"entry": {"name": "pointer-name", "number": 0}}
+>>> default('{}')
+{"entry": {"name": "pointer-name", "number": 0}}
+>>> default('{entry: null}')
+{"entry": {"name": "pointer-name", "number": 0}}
+>>> default('{entry: {}}')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{entry: {name: other-name}}')
+{"entry": {"name": "other-name", "number": 0}}
+```
+
+#### Non-pointer fields
+
+Non-pointer fields are also important, since they have a specific unmarshalling mechanism in Go.
+
+For CRDs, we can assume that invalid zero-values are omitted, otherwise
+the objects would technically be invalid.
+
+For built-ins, we can differentiate two cases:
+1. The zero-value for the field is invalid (e.g. if empty string is not a valid value)
+2. The zero-value is valid, but is also the default (e.g. enabled false by default).
+
+Both these cases will be handled properly, here's an example:
+```golang
+type Object struct {
+     // This field has an invalid zero-value, empty string is not a
+     // valid string.
+     // +default=default-name
+     Name string
+
+     // This field is defaulted to 0 with no defaulter.
+     // +default=0
+     Defaulted int
+}
+```
+
+This would generate the following OpenAPI:
+```yaml
+Object:
+  type: object
+  properties:
+    invalid:
+      type: string
+      default: default-name
```

Comment:
why invalid zero value? Where is it a) in the type above and b) here in OpenAPI.

Reply:
With this OpenAPI spec `""` is a completely valid value.